### PR TITLE
[CSL-2113] Binance new address performance improvement

### PR DESCRIPTION
**Pull request by pva701**
There are two improvements:
1. Checking existence of an account in `newAddress` is done much faster. Previously it was done by calling `getAccount`, which collects all addresses, deserialize them and it's extremely slow.
2. Generating unique address is optimized. For each generated address we derived it even this address already existed in db or was hardened, now we check it at first and then derive

I haven't generated 80k addresses like binance contains, but for 1k addresses, the first improvement showed much better performance

### CODE REVIEW

File: wallet/src/Pos/Wallet/Web/Account.hs [hunk 1]
```diff
@@ -132,14 +132,13 @@ generateUnique desc RandomSeed generator isDuplicate = loop (100 :: Int)
...
         rand  <- liftIO randomIO
-        value <- generator rand
-        bad   <- orM
-            [ isDuplicate rand value
-            , pure $ isHardened rand  -- using hardened keys only for now
-            ]
-        if bad
-            then loop (i - 1)
-            else return value
+        if isHardened rand then
```
Comment (et4te) [inline wallet/src/Pos/Wallet/Web/Account.hs]: Shouldn't this be `if isHardened rand then ... else loop (i - 1)` ? This change afaik was merged previously here: https://github.com/input-output-hk/cardano-sl/pull/1991/commits/4ebc883d4f08ffa13d3bce74e71792a3a54aeb42.

Comment (pva701) [reply wallet/src/Pos/Wallet/Web/Account.hs]: @et4te as I can see we considered index which `>= firstHardened` as bad and called loop again,  but your commit considers indexes `>= firstHardened` as good. afaiu it was done intentionally, but why?

Comment (adinapoli-iohk) [reply wallet/src/Pos/Wallet/Web/Account.hs]: @pva701 maybe I'm missing something, but requirements dictates we need to generate only hardened keys for addresses, for security reasons. See also here for a comment from [Dmitry](https://github.com/input-output-hk/cardano-sl/pull/1991#issuecomment-347971486), explicitly stating that what Edward fixed was a genuine bug.

File: wallet/src/Pos/Wallet/Web/Account.hs [hunk 2]
```diff
@@ -132,14 +132,13 @@ generateUnique desc RandomSeed generator isDuplicate = loop (100 :: Int)
...
+            loop (i - 1) -- use non-hardened keys only for now
+        else do
+            value <- generator rand
+            isDup <- isDuplicate rand value
```
Comment (adinapoli-iohk) [inline wallet/src/Pos/Wallet/Web/Account.hs]: @pva701 Have you tried speeding up the underlying function which `isDuplicate` uses as described [here](https://iohk.myjetbrains.com/youtrack/issue/CSL-2120#comment=93-16616)? Do you notice any substantial improvement or was I being mental there?

**Comment by adinapoli-iohk** [comment]
@dcoutts I finally had the time to put [db-gen](https://github.com/adinapoli-iohk/dbgen) to good use in generating a DB with 20k addresses, and then benchmarked @pva701 's patch. Here are the results in generating 10 addresses:

## DB Stats

```
dbgen [master] ⚡ stack exec dbgen -- --nodeDB ../cardano-sl/run/node-db0 --walletDB wallet-db --add-to Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469 --stats
Wallets: 1

- Wallet #1, ready, Synced[AbstractHash 6e5377d2c78b8ba69644bc17caa8c05131d9dd32ed95e8da7a2c3babcad90e6e], PendingTxs: 0


Accounts: 2

- Initial account, Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@2147483648, addresses: 1, removed: 0
- Account Number #1, Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469, addresses: 20309, removed: 0


Number of used addresses: 0
Number of change addresses: 0
```

## unpatched

```
dbgen [master] stack exec dbgen -- --nodeDB ../cardano-sl/run/node-db0 --walletDB wallet-db --add-to Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469
Faking StateLock syncing...
Generating 10 addresses for Account Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469...
Action took 26.411885 seconds.
OK.
```

## Ilya's patch

```
dbgen [master] ⚡ stack exec dbgen -- --nodeDB ../cardano-sl/run/node-db0 --walletDB wallet-db --add-to Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469 +RTS -p
Faking StateLock syncing...
Generating 10 addresses for Account Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469...
Action took 0.47962 seconds.
OK.
```

I have also tried the code I proposed (aka rewriting the `doesWAddressExist` for good) but that yielded marginal if not negligible speedup, so the culprit was effectively (as per Ilya's description) the deserialisation cost which was eating all the time. Here is the summary of the `.prof` file generated by the latest run (featuring mine and Ilya's work all in one patch):

```
	Tue Jan  2 12:48 2018 Time and Allocation Profiling Report  (Final)

	   dbgen +RTS -N -p -RTS --nodeDB ../cardano-sl/run/node-db0 --walletDB wallet-db --add-to Ae2tdPwUPEZGwUVR8meJ6mgbHeiC16TNyM3wTDYRYwDBMeCHUXdMVLHKupC@75378469

	total time  =        3.07 secs   (3071 ticks @ 1000 us, 1 processor)
	total alloc = 2,722,793,696 bytes  (excludes profiling overheads)

COST CENTRE                   MODULE                          SRC                                                   %time %alloc

scrypt.\.\.\                  Crypto.Scrypt                   src/Crypto/Scrypt.hs:(244,56)-(250,66)                 27.7    0.0
decodeWithTable/fill          Data.ByteString.Base64.Internal Data/ByteString/Base64/Internal.hs:(171,62)-(194,68)    7.4   15.1
>>=.\.\                       Data.Serialize.Get              src/Data/Serialize/Get.hs:222:49-74                     7.1    2.2
decodeWithTable.\.\.look      Data.ByteString.Base64.Internal Data/ByteString/Base64/Internal.hs:(165,11)-(168,49)    4.9   18.6
crc16                         Data.Acid.CRC                   src/Data/Acid/CRC.hs:(55,1)-(57,76)                     4.0   10.8
crc16.worker                  Data.Acid.CRC                   src/Data/Acid/CRC.hs:(56,11)-(57,76)                    3.7    3.6
>>=.\                         Data.Serialize.Get              src/Data/Serialize/Get.hs:222:7-74                      2.2    1.5
normalizePositive             Data.Scientific                 src/Data/Scientific.hs:(1035,1)-(1038,61)               1.9    2.9
parserParseOne'.\             Text.Libyaml                    Text/Libyaml.hs:(562,57)-(574,34)                       1.5    0.1
>>=.\.ks'                     Data.Aeson.Types.Internal       Data/Aeson/Types/Internal.hs:270:43-76                  1.2    0.3
>>=.\                         Data.Aeson.Types.Internal       Data/Aeson/Types/Internal.hs:(270,39)-(271,65)          1.2    1.1
fromAvvmPk                    Pos.Crypto.Signing.Types.Redeem Pos/Crypto/Signing/Types/Redeem.hs:(88,1)-(97,35)       1.1    1.7
textToValue.isLike            Data.Yaml.Internal              Data/Yaml/Internal.hs:(175,9)-(176,69)                  1.1    2.0
replace.x.loop                Data.Text                       Data/Text.hs:(708,11)-(714,64)                          1.1    0.0
decodeWithTable.\.\.fill      Data.ByteString.Base64.Internal Data/ByteString/Base64/Internal.hs:(169,11)-(194,68)    1.0    1.9
liftParseJSON.\               Data.Aeson.Types.FromJSON       Data/Aeson/Types/FromJSON.hs:1536:39-88                 0.9    2.7
textToValue.isLike.titleCased Data.Yaml.Internal              Data/Yaml/Internal.hs:176:17-69                         0.5    1.8
encodeUtf8                    Data.Text.Encoding              Data/Text/Encoding.hs:(373,1)-(388,37)                  0.3    1.3
memXor                        Data.Memory.PtrMethods          Data/Memory/PtrMethods.hs:(42,1)-(45,66)                0.3    2.3
```

TL;DR Ilya's work should yield a sensible speedup also for large datasets. cc @domenkozar

**Comment by domenkozar** [comment]
@adinapoli-iohk @dcoutts once you approve the PR I can ship this improvement from master.

File: wallet/src/Pos/Wallet/Web/Account.hs (hunk 1 above)
Comment (kantp) [reply wallet/src/Pos/Wallet/Web/Account.hs]: But unless I'm missing something, merging this as is would undo the bugfix in #1991 -- surely we do not want that?

I agree with @et4te that we should preserve the fix of 4ebc883, at least before deploying this. @adinapoli-iohk, you did benchmark the improvements without the bugfix already, so we do know that we have a performance improvement, and need not worry about it being masked, right?

Comment (kantp) [reply wallet/src/Pos/Wallet/Web/Account.hs]: I'd like this to preserve the bugfix of #1991

**Comment by pva701** [comment]
@kantp cherry-picked #1991

**Comment by adinapoli-iohk** [comment]
> once you approve the PR I can ship this improvement from master.

Go ahead, @domenkozar !

File: wallet/src/Pos/Wallet/Web/Account.hs (hunk 1 above)
Comment (adinapoli-iohk) [reply wallet/src/Pos/Wallet/Web/Account.hs]: > so we do know that we have a performance improvement, and need not worry about it being masked, right?

Absolutely correct, Philipp, that's why I'm happy now 😉
